File: learning-servers/nim/src/validate.nim
import std/[strutils, json]
import mummy

proc isCountry(s: string): bool =
  s in ["US", "CA", "GB", "DE", "FR", "JP", "AU", "BR", "IN", "CN"]

proc validUsername(s: string): bool =
  if s.len < 3 or s.len > 30: return false
  for ch in s:
    if not ((ch >= 'a' and ch <= 'z') or (ch >= '0' and ch <= '9') or ch == '_'): return false
  return true

proc validEmail(s: string): bool =
  if s.len == 0 or s.len > 100 or ' ' in s: return false
  let at = s.find('@')
  if at <= 0: return false
  let domain = s[at + 1 .. ^1]
  if '@' in domain: return false
  let dot = domain.find('.')
  return dot > 0 and dot != domain.len - 1

proc validPassword(s: string): bool =
  if s.len < 8 or s.len > 100: return false
  var lo, up, di = false
  for ch in s:
    if ch >= 'a' and ch <= 'z': lo = true
    elif ch >= 'A' and ch <= 'Z': up = true
    elif ch >= '0' and ch <= '9': di = true
  return lo and up and di

proc validSku(s: string): bool =
  if s.len != 7 or s[3] != '-': return false
  for i in 0 .. 2:
    if s[i] < 'A' or s[i] > 'Z': return false
  for i in 4 .. 6:
    if s[i] < '0' or s[i] > '9': return false
  return true

proc validatePayload(jsonBody: string): bool =
  var node: JsonNode
  try:
    node = parseJson(jsonBody)
  except CatchableError:
    return false
  if node.kind != JObject or node.len != 8: return false
  for k in ["username", "email", "age", "password", "website", "country", "tags", "items"]:
    if not node.hasKey(k): return false
  if node["username"].kind != JString or not validUsername(node["username"].getStr): return false
  if node["email"].kind != JString or not validEmail(node["email"].getStr): return false
  if node["age"].kind != JInt: return false
  let age = node["age"].getInt
  if age < 13 or age > 120: return false
  if node["password"].kind != JString or not validPassword(node["password"].getStr): return false
  if node["website"].kind != JString: return false
  let web = node["website"].getStr
  if web.len > 200 or not (web.startsWith("http://") or web.startsWith("https://")): return false
  if node["country"].kind != JString or not isCountry(node["country"].getStr): return false
  let tags = node["tags"]
  if tags.kind != JArray or tags.len < 1 or tags.len > 10: return false
  for t in tags:
    if t.kind != JString or t.getStr.len < 1 or t.getStr.len > 20: return false
  let items = node["items"]
  if items.kind != JArray or items.len < 1 or items.len > 50: return false
  for it in items:
    if it.kind != JObject or it.len != 3: return false
    if not (it.hasKey("sku") and it.hasKey("qty") and it.hasKey("price")): return false
    if it["sku"].kind != JString or not validSku(it["sku"].getStr): return false
    if it["qty"].kind != JInt: return false
    let q = it["qty"].getInt
    if q < 1 or q > 999: return false
    var price: float
    case it["price"].kind
    of JFloat: price = it["price"].getFloat
    of JInt: price = it["price"].getInt.float
    else: return false
    if price < 0 or price > 100000: return false
  return true

proc validate*(request: Request) =
  var headers: HttpHeaders
  headers["Content-Type"] = "application/json"
  if validatePayload(request.body):
    request.respond(200, headers, "{\"valid\":true}")
  else:
    request.respond(400, headers, "{\"valid\":false}")

File: learning-servers/zig/src/validate.zig
const std = @import("std");

const zap = @import("zap");

const Item = struct { sku: []const u8, qty: i64, price: f64 };
const Payload = struct {
    username: []const u8,
    email: []const u8,
    age: i64,
    password: []const u8,
    website: []const u8,
    country: []const u8,
    tags: [][]const u8,
    items: []Item,
};

fn validUsername(s: []const u8) bool {
    if (s.len < 3 or s.len > 30) return false;
    for (s) |ch| {
        const ok = (ch >= 'a' and ch <= 'z') or (ch >= '0' and ch <= '9') or ch == '_';
        if (!ok) return false;
    }
    return true;
}

fn validEmail(s: []const u8) bool {
    if (s.len == 0 or s.len > 100) return false;
    for (s) |ch| if (ch == ' ') return false;
    const at = std.mem.indexOfScalar(u8, s, '@') orelse return false;
    if (at == 0) return false;
    const domain = s[at + 1 ..];
    if (std.mem.indexOfScalar(u8, domain, '@') != null) return false;
    const dot = std.mem.indexOfScalar(u8, domain, '.') orelse return false;
    return dot != 0 and dot != domain.len - 1;
}

fn validPassword(s: []const u8) bool {
    if (s.len < 8 or s.len > 100) return false;
    var lo = false;
    var up = false;
    var di = false;
    for (s) |ch| {
        if (ch >= 'a' and ch <= 'z') {
            lo = true;
        } else if (ch >= 'A' and ch <= 'Z') {
            up = true;
        } else if (ch >= '0' and ch <= '9') {
            di = true;
        }
    }
    return lo and up and di;
}

fn validSku(s: []const u8) bool {
    if (s.len != 7 or s[3] != '-') return false;
    for (s[0..3]) |ch| if (ch < 'A' or ch > 'Z') return false;
    for (s[4..7]) |ch| if (ch < '0' or ch > '9') return false;
    return true;
}

fn validCountry(s: []const u8) bool {
    const set = [_][]const u8{ "US", "CA", "GB", "DE", "FR", "JP", "AU", "BR", "IN", "CN" };
    for (set) |x| if (std.mem.eql(u8, x, s)) return true;
    return false;
}

fn validatePayload(p: Payload) bool {
    if (!validUsername(p.username)) return false;
    if (!validEmail(p.email)) return false;
    if (p.age < 13 or p.age > 120) return false;
    if (!validPassword(p.password)) return false;
    if (p.website.len > 200) return false;
    if (!std.mem.startsWith(u8, p.website, "http://") and !std.mem.startsWith(u8, p.website, "https://")) return false;
    if (!validCountry(p.country)) return false;
    if (p.tags.len < 1 or p.tags.len > 10) return false;
    for (p.tags) |t| if (t.len < 1 or t.len > 20) return false;
    if (p.items.len < 1 or p.items.len > 50) return false;
    for (p.items) |it| {
        if (!validSku(it.sku)) return false;
        if (it.qty < 1 or it.qty > 999) return false;
        if (it.price < 0 or it.price > 100000) return false;
    }
    return true;
}

pub fn validate(r: zap.Request) !void {
    r.setHeader("content-type", "application/json") catch {};
    const req_body = r.body orelse {
        r.setStatus(.bad_request);
        try r.sendBody("{\"valid\":false}");
        return;
    };
    const parsed = std.json.parseFromSlice(Payload, std.heap.c_allocator, req_body, .{}) catch {
        r.setStatus(.bad_request);
        try r.sendBody("{\"valid\":false}");
        return;
    };
    defer parsed.deinit();
    if (validatePayload(parsed.value)) {
        try r.sendBody("{\"valid\":true}");
    } else {
        r.setStatus(.bad_request);
        try r.sendBody("{\"valid\":false}");
    }
}
